### ISSUE TITLE
Rootless Kubeflow

### ISSUE THREAD

**Issue by juliusvonkohout** [kubeflow/manifests#2528]
# Rootless Kubeflow

This is an issue to track the progress of the following proposal https://github.com/kubeflow/manifests/blob/master/proposals/20200913-rootlessKubeflow.md.
### Goals

We want to run Kubeflow as rootless as possible according to CNCF/Kubernetes best practices.
Most enterprise environments will require this as well.

### Implementation details
The main steps are adding an additional profile for istio-cni and later ambient mesh, updating the documentation and manifest generation process.
Only istio-cni or istio ambient mesh can run rootless as explained here https://istio.io/latest/docs/setup/additional-setup/cni/. 
Istio-cni will still nedd a deamonset in kube-system, but that is completly isolated from user workloads. 
The ambient mesh should get rid of this as well and also has the benefit of removing the istio initcontainers and sidecars altogether.
Then adding the baseline and restricted PSS as kustomize component to /contrib and extending the profile controller to annotate user namespaces with configurable PSS labels.

We want to use a staged approach.

First Stage:
1. Implement Istio 1.17.5 and use it by default, because 1.17. is what we have planned to use for Kubeflow 1.8.
2. Implement istio-cni (--set components.cni.enabled=true --set components.cni.namespace=kube-system) as second option.
3. Add simple tests similar to tests/gh-actions/install_istio.sh and tests/gh-actions/install_knative.sh for istio-cni and support both rootfull and rootless istio at the same time and give users one release to test

Second stage:
4. Add pod security standards (https://kubernetes.io/docs/concepts/security/pod-security-standards/) base/restricted to manifests/contrib
5. Enforce PSS baseline (here you can still build OCI containers via Podman and buildah). It works with any istio
6. Enable Warnings for violations of restricted PSS
7. Add tests to make sure that the PSS are used and tested in the CICD
8. Optionally Enforce PSS restricted (this is where minor corner cases are affected)

Third stage:
9. Upgrade Istio to 1.19 to make the ambient mesh available
10. Add istio-ambient as an option to the next Kubeflow release.

fourth stage:
11. Use the ambient service mesh by default in Kubeflow 1.10.

### Non-Goals
This does not cover Application level CVEs, only cluster level security.

### Does this break any existing functionality?
So far not. Only PSS restricted may block the incredibly dangerous and unprofessional Docker in Docker.
This is a rarely used feature from the KFP SDK.
With PSS baseline you can still build OCI images with Podman for example. 
We should replace Docker with the cli compatible podman in the KFP SDK https://kubeflow-pipelines.readthedocs.io/en/1.8.22/source/kfp.containers.html?highlight=kfp.containers.build_image_from_working_dir#kfp.containers.build_image_from_working_dir.

### Does this fix/solve any outstanding issues?
We are not following best practices and this is forbidden in most enterprise environments.
The progress is tracked in https://github.com/kubeflow/manifests/issues/2528

**Comment by juliusvonkohout** [comment]
The first stage has been implemented in https://github.com/kubeflow/manifests/pull/2455

**Comment by juliusvonkohout** [comment]
superseedes https://github.com/kubeflow/manifests/issues/2014

**Comment by peterj** [comment]
I have experience with Istio/Ambient mesh and would love to help out if kubeflow is planning to adopt ambient in the future.

**Comment by juliusvonkohout** [comment]
@peterj please reach out on the Kubeflow Slack or Linkedin.

**Comment by diegolovison** [comment]
Should we have also a definition of done for this issue being able to deploy using `kind` and `podman`?

**Comment by juliusvonkohout** [comment]
https://github.com/kubeflow/manifests/pull/2714 is related

**Comment by juliusvonkohout** [comment]
@biswajit-9776 is working on this issue

**Comment by sridhar1029** [comment]
Hi everyone,

I'm working on improving the security of our Kubeflow deployment and are exploring the possibility of running notebook server containers as non-root users (e.g., with runAsUser: 1000, runAsNonRoot: true, and allowPrivilegeEscalation: false).

However, many of the default Kubeflow notebook images seem to assume root access, and attempting to enforce non-root policies (e.g., via Kyverno) results in startup failures, particularly due to issues like:
`/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 1000...`

👉 Is there a supported or recommended approach for running notebook servers as non-root in Kubeflow?
- Are any of the official notebook images compatible with non-root operation?
- If not, what changes would be needed in the Dockerfile or entrypoint to support this?
- Is there any roadmap consideration for shipping hardened/non-root images?
- Any guidance or examples would be appreciated!

Thanks!

**Comment by juliusvonkohout** [comment]
> Hi everyone,
> 
> I'm working on improving the security of our Kubeflow deployment and are exploring the possibility of running notebook server containers as non-root users (e.g., with runAsUser: 1000, runAsNonRoot: true, and allowPrivilegeEscalation: false).
> 
> However, many of the default Kubeflow notebook images seem to assume root access, and attempting to enforce non-root policies (e.g., via Kyverno) results in startup failures, particularly due to issues like: `/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 1000...`
> 
> 👉 Is there a supported or recommended approach for running notebook servers as non-root in Kubeflow?
> 
>     * Are any of the official notebook images compatible with non-root operation?
> 
>     * If not, what changes would be needed in the Dockerfile or entrypoint to support this?
> 
>     * Is there any roadmap consideration for shipping hardened/non-root images?
> 
>     * Any guidance or examples would be appreciated!
> 
> 
> Thanks!

Our CI/CD in the master branch already uses PSS restricted for static and PSS baseline for dynamic namespaces. Please start with a fresh cluster and the master branch to test it.

**Comment by sridhar1029** [comment]
Hi @juliusvonkohout,

Thanks for the quick response.

I’ve tested the latest master branch, but I’m still running into the same issue. I’m setting up a multi-tenant Kubeflow cluster and would like to enforce PSS `restricted` in the dynamically created user profile namespaces to improve security isolation between users.

However, in the current state, this doesn’t seem to work well, as several Kubeflow components still require running as root — which violates the restricted PSS policy.

Is there any ongoing work or roadmap around making Kubeflow more compatible with restricted PSS in user namespaces? Or is the only path forward to rebuild the notebook images and possibly patch some controllers to make them PSS-compliant?

Another alternative could be to upgrade the cluster from Kubernetes 1.31 to 1.33 to take advantage of the user namespaces?

Thanks for your help!

**Comment by juliusvonkohout** [comment]
> Hi [@juliusvonkohout](https://github.com/juliusvonkohout),
> 
> Thanks for the quick response.
> 
> I’ve tested the latest master branch, but I’m still running into the same issue. I’m setting up a multi-tenant Kubeflow cluster and would like to enforce PSS `restricted` in the dynamically created user profile namespaces to improve security isolation between users.
> 
> However, in the current state, this doesn’t seem to work well, as several Kubeflow components still require running as root — which violates the restricted PSS policy.
> 
> Is there any ongoing work or roadmap around making Kubeflow more compatible with restricted PSS in user namespaces? Or is the only path forward to rebuild the notebook images and possibly patch some controllers to make them PSS-compliant?
> 
> Another alternative could be to upgrade the cluster from Kubernetes 1.31 to 1.33 to take advantage of the user namespaces?
> 
> Thanks for your help!

Yes and you can help. By default we enforce PSS baseline for dynamic namespaces on the master branch. Now if you switch from Minio to seweadfs as used in the kubeflow/manifests CI/Cd, then the basic usernamespaces should work with PSS. but for all user workloads you add you have to make sure that you set a proper securitycontext.

**Comment by juliusvonkohout** [comment]
So please try it out with structured feedback and analysis and lets fix the remaining items.

**Comment by sridhar1029** [comment]
Hi @juliusvonkohout ,

Thanks for the suggestion. I tried switching to SeaweedFS as recommended, but I’ve run into a few issues that I’m currently looking into:

- **PSS Enforcement**: I don’t see the PSS baseline being applied to dynamic namespaces (user profiles) on the latest master branch. New profiles are created without any PSS-related labels.

- **SeaweedFS Compatibility**: SeaweedFS doesn’t seem to resolve the compatibility issues with PSS: restricted. I'm still encountering errors when trying to apply non-root security contexts to notebook pods. These pods attempt to write to the root filesystem, which I’d like to avoid in user namespaces for security reasons.

- **Pipelines**: The pipelines seem to be broken with SeaweedFS. They still attempt to use the old MinIO secrets, which causes failures during execution.

I’ll continue digging into the SeaweedFS configuration and the overall behaviour with user namespaces. If there are any known compatibility limitations or suggested debug steps, please let me know — it would really help narrow things down faster.

**Comment by juliusvonkohout** [comment]
> Hi [@juliusvonkohout](https://github.com/juliusvonkohout) ,
> 
> Thanks for the suggestion. I tried switching to SeaweedFS as recommended, but I’ve run into a few issues that I’m currently looking into:
> 
>     * **PSS Enforcement**: I don’t see the PSS baseline being applied to dynamic namespaces (user profiles) on the latest master branch. New profiles are created without any PSS-related labels.
> 
>     * **SeaweedFS Compatibility**: SeaweedFS doesn’t seem to resolve the compatibility issues with PSS: restricted. I'm still encountering errors when trying to apply non-root security contexts to notebook pods. These pods attempt to write to the root filesystem, which I’d like to avoid in user namespaces for security reasons.
> 
>     * **Pipelines**: The pipelines seem to be broken with SeaweedFS. They still attempt to use the old MinIO secrets, which causes failures during execution.
> 
> 
> I’ll continue digging into the SeaweedFS configuration and the overall behaviour with user namespaces. If there are any known compatibility limitations or suggested debug steps, please let me know — it would really help narrow things down faster.

We actually test this live in our CI/CD here is the proof with PSS tests and seaweedfs and testing all components basic functionality https://github.com/kubeflow/manifests/actions/runs/16771055044/job/47486020364 and https://github.com/kubeflow/manifests/pull/3204 was also merged. 

<img width="530" height="236" alt="Image" src="https://github.com/user-attachments/assets/bb595e76-0991-4ce7-b847-71418a291056" />

So you only need to use the latest master and check whether https://github.com/kubeflow/manifests/pull/3204 is applied properly for the user namespaces via https://github.com/kubeflow/manifests/blob/97ade2830c88b568d59a8ea19a1b1a27a76748a5/example/kustomization.yaml#L110

**Comment by sridhar1029** [comment]
Yeah, the static namespaces in my deployment were also labeled correctly. I was under the impression that newly created profiles (dynamic user namespaces) would have the PSS baseline enforced by default. Is that not the case?

**Comment by juliusvonkohout** [comment]
> Yeah, the static namespaces in my deployment were also labeled correctly. I was under the impression that newly created profiles (dynamic user namespaces) would have the PSS baseline enforced by default. Is that not the case?

Did you thoroughly read all the details of my comment above https://github.com/kubeflow/manifests/issues/2528#issuecomment-3166964837 ?

**Comment by sridhar1029** [comment]
Sorry, will fetch the latest

**Comment by sridhar1029** [comment]
Sorry for the confusion. I was installing individual components, and I think I forgot to apply `/common/security/PSS/dynamic/baseline`.

**Comment by juliusvonkohout** [comment]
It is almost done in the master branch. Lets evaluate for Kubeflow 1.11

**Comment by juliusvonkohout** [comment]
It is mostly done in Kubeflow 1.11 now we need to tackle the customer namespaces to get there from baseline to pss restricted. probably by fixing the securitycontext of KFP and jupyterlabs.

**Comment by juliusvonkohout** [comment]
It mostly done now given @jsonmp-k8 and my changes in KFP 0.16.0 around podsecuritystandards. I am just thinking about enforcing runasnonroot by default for KFP and enforce pss restricted instead of baseline by default.

**Comment by Ahad-Dngwala** [comment]
Hi @juliusvonkohout ,

I’ve been reading through the **Rootless Kubeflow** discussion and the recent updates around Pod Security Standards and `runAsNonRoot` in KFP. The work on moving from **baseline → restricted PSS** looks really interesting.

I’d like to help with this effort if there are still pieces that need attention. I’m currently going through the manifests and trying to understand which components (for example KFP or Jupyter notebooks) might still require adjustments in their `securityContext` to work properly under restricted PSS.

If there’s a particular component, manifest, or test area that would be useful to start with, I’d be happy to take a look and propose a small fix or improvement.

**Comment by juliusvonkohout** [comment]
https://github.com/kubeflow/manifests/pull/3412 might be the last thing needed, so I am closing the issue.
